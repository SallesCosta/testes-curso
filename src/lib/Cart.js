import find from 'lodash/find'
import remove from 'lodash/remove'
import Dinero from 'dinero.js'

const Money = Dinero

const calculateDiscount = (amount, quantity, condition) => {
  const list = Array.isArray(condition) ? condition : [condition]

  const [higherDiscount] = list
    .map(cond => {
      if (cond.percentage) {
        return calculatePercentageDiscount(amount, {
          condition: cond,
          quantity,
        }).getAmount()
      } else if (cond.quantity) {
        return calculateQuantityDiscount(amount, {
          condition: cond,
          quantity,
        }).getAmount()
      }
    })
    .sort((a, b) => b - a)

  return Money({amount: higherDiscount})
}

const calculatePercentageDiscount = (amount, item) => {
  if (item.condition?.percentage && item.quantity > item.condition.minimum) {
    return amount.percentage(item.condition.percentage)
  }
  return Money({ amount: 0 })
}

const calculateQuantityDiscount = (amount, item) => {
  const isEven = item.quantity % 2 === 0

  if (item.condition?.quantity && item.quantity > item.condition.quantity) {
    return amount.percentage(isEven ? 50 : 40)
  }
  return Money({ amount: 0 })
}

Money.defaultCurrency = 'BRL'
Money.defaultPrecision = 2

export default class Cart {
  items = []

  getTotal() {
    return this.items.reduce((acc, item) => {
      const amount = Money({ amount: item.quantity * item.product.price })

      let discount = Money({ amount: 0 })

      if (item.condition) {
        discount = calculateDiscount(amount, item.quantity, item.condition)
      }
      // if (item.condition?.percentage) {
      //   discount = calculatePercentageDiscount(amount, item)
      // } else if (item.condition?.quantity) {
      //   discount = calculateQuantityDiscount(amount, item)
      // }

      return acc.add(amount).subtract(discount)
    }, Money({ amount: 0 }))
  }

  add(item) {
    const itemToFind = { product: item.product }
    if (find(this.items, itemToFind)) {
      remove(this.items, itemToFind)
    }
    this.items.push(item)
  }

  remove(product) {
    remove(this.items, { product })
  }

  summary() {
    const total = this.getTotal().getAmount()
    const items = this.items

    return {
      total,
      items,
    }
  }

  checkout() {
    const { total, items } = this.summary()

    this.items = []

    return {
      total,
      items,
    }
  }
}

const cart = new Cart()

const product = {
  title: '',
  price: '',
}

const item = {
  quantity: 2,
  product,
}

cart.add(item)
cart.remove(product)
cart.getTotal()
cart.summary()
cart.checkout()
