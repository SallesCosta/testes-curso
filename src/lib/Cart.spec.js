import Cart from './Cart'
describe('Cart', () => {
  let cart

  let product = {
    title: 'Adidas running shoes - men',
    price: 35388,
  }

  let product2 = {
    title: 'Adidas running shoes - women',
    price: 41872,
  }

  beforeEach(() => {
    cart = new Cart()
  })

  describe('gettotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toEqual(0)
    })

    it('should multiply quantity and price and recive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      }
      cart.add(item)
      expect(cart.getTotal().getAmount()).toEqual(70776)
    })

    it('should ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product,
        quantity: 1,
      })

      expect(cart.getTotal().getAmount()).toEqual(35388)
    })

    it('shoul update total when a product gets included and then remove', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 1, // 112.648
      })

      cart.remove(product)
      expect(cart.getTotal().getAmount()).toEqual(41872)
    })
  })

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 3,
      })

      expect(cart.checkout()).toMatchSnapshot()
    })

    it('should reset the cart when checktou() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      })
      cart.checkout()
      expect(cart.getTotal().getAmount()).toEqual(0)
    })

    it('should return an object with the total and the list fo items when summary() is called)', async () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 3,
      })

      expect(cart.summary()).toMatchSnapshot()
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0)
    })
  })

  describe('special conditions', () => {
    it('should NOT apply percentage discount quantity is below or equals minimum', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 2,
      })

      expect(cart.getTotal().getAmount()).toEqual(70776)
    })

    it('should apply percentage discount quantity above minimum in passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 3,
      })

      expect(cart.getTotal().getAmount()).toEqual(74315)
    })

    it('should aplly quantity discount for even quantities when condition in no met', async () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 1,
      })

      expect(cart.getTotal().getAmount()).toEqual(35388)
    })

    it('should aplly quantity discount for even quantities', async () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 4,
      })

      expect(cart.getTotal().getAmount()).toEqual(70776)
    })

    it('should aplly quantity discount for odd quantities', async () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 5,
      })

      expect(cart.getTotal().getAmount()).toEqual(106164)
    })

    it('should receive two or more conditions and determine/apply the best discount. First case', async () => {
      const condition1 = {
        percentage: 30,
        minimum: 2,
      }
      const condition2 = {
        quantity: 2,
      }
      cart.add({
        product,
        consition: [condition1, condition2],
        quaitty: 5,
      })
      expect(cart.getTotal().getAmount().toEqual(106164))
    })
  })
})
