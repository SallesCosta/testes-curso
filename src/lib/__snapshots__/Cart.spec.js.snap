// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should reset the cart when checktou() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 125616,
}
`;

exports[`Cart checkout() should return an object with the total and the list fo items when summary() is called) 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 3,
    },
  ],
  "total": 196392,
}
`;
